### DISCUSSION TITLE
Fallback to polling does not work

### DISCUSSION THREAD

**Discussion by RonnieRocket147** [socketio/socket.io#3998] [Q&A] [answered]
We have a couple of customers who are unable to connect with the server due to the websocket being blocked by their network. Our client transports are configured as:

```["websocket", "polling"]```

I would suspect the client to fallback to polling however this does not work in their situation. When the following transports order is used the polling connection IS working for them:

```["polling", "websocket"]```

See the image of what the customer is seeing in the browser console when the websocket is specified as first option.

Why is the fallback not working in this case?

![MicrosoftTeams-image](https://user-images.githubusercontent.com/28155950/124169375-78bcd680-daa6-11eb-820c-381d6b0e0993.png)

**Comment by darrachequesne** [comment, marked as answer]
Hi! I think the word "fallback" may be a bit misleading here, as the order of the transports in the `transports` option matters:

- `["websocket", "polling"]` means: establish a WebSocket connection first (or use HTTP long-polling if WebSocket is not available, for example in older browsers)
- `["polling", "websocket"]` means: establish a HTTP long-polling connection first, and then upgrade to WebSocket

Related: https://github.com/socketio/engine.io-client/issues/575

**Reply by RonnieRocket147** [reply]
Hmm interesting!

```["websocket", "polling"]```: so in a modern browser polling will never be used even if the Websocket connection cannot be established (like in my example above)?

Is there a scenario possible like _try Websocket first, if not possible then use polling_? I think that would be better than poll-and-upgrade-if-possible nowadays.

Thanks again!

**Reply by darrachequesne** [reply]
> ["websocket", "polling"]: so in a modern browser polling will never be used even if the Websocket connection cannot be established (like in my example above)?

Yes, that's right, unfortunately.

> Is there a scenario possible like try Websocket first, if not possible then use polling? I think that would be better than poll-and-upgrade-if-possible nowadays.

I think the easiest way to do this is to catch the  `connect_error` event:

```js
const socket = io({
  transports: ["websocket", "polling"]
});

socket.on("connect_error", () => {
  // revert to classic upgrade
  socket.io.opts.transports = ["polling", "websocket"];
});
```

Reference: https://socket.io/docs/v4/client-socket-instance/#connect-error

**Reply by RonnieRocket147** [reply]
Thanks Damien,

That looks like it could work, I'll give that a try.

Many thanks for taking the time to respond.

**Reply by darrachequesne** [reply]
You're welcome! I updated the documentation here: https://socket.io/docs/v4/client-initialization/#transports
